src/sentry/snuba/metrics.py

```diff
@@ -118,13 +119,20 @@ def __init__(self, query_params, allow_minute_resolution=False):
         self.start = start
         self.end = end
 
-    def get_intervals(self):
-        start = self.start
-        end = self.end
-        delta = timedelta(seconds=self.rollup)
-        while start < end:
-            yield start
-            start += delta
+
+class TimeRange(Protocol):
+    start: datetime
+    end: datetime
+    rollup: int
+
+
+def get_intervals(query: TimeRange):
```

Comment:
can't you use the real type here? I think this protocol may be too lax

Reply:
This function is used with multiple types. What do mean by too lax? The protocol defines exactly what this function needs.

Reply:
ok then a protocol is alright. I mean the types can mean diff things. just because a type has a property `foo: int` doesn't mean `foo` means the same thing and is in the same unit every time. Since TimeRange is a small protocol it's easy to impl on accident -> more types are accepted on accident

but I think this is fine, the field names are somewhat specific/descriptive so I see no risk